### PR TITLE
Optimizer class

### PR DESCRIPTION
**Pull request by WanderingMike**
This is a pull request for a new Optimizer class, for the sole purpose of discussing the design choices and not for merging. For now, only the SGD and the CMA are properly implemented. However, for the SGD the gradient methods of another pull request are necessary to run it.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

File: src/qibo/derivative.py [hunk 1]
```diff
@@ -1,7 +1,180 @@
...
+        return self._apply_func(fixed_params=params)
+
+    def get_scaling_factor(self, trainablep_idx):
+        """Get scaling factor multiplying a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = 1.0
+        return self._apply_func(fixed_params=params)
+
+
+def create_hamiltonian(qubit, nqubit, backend):
```
Comment (MatteoRobbiati) [inline src/qibo/derivative.py:64]: I think this kind of function should be defined by the user. We should provide a general method which takes as argument a general Hamiltonian.

File: src/qibo/derivative.py [hunk 2]
```diff
@@ -1,7 +1,180 @@
...
+            hams.append(I(i))
+
+    # create Hamiltonian
+    obs = np.prod([Z(i) for i in range(1)])
+    hamiltonian = SymbolicHamiltonian(obs, backend=backend)
+
+    return hamiltonian
+
+
+def error_mitigation(circuit, hamiltonian, backend, noise_model):
```
Comment (MatteoRobbiati) [inline src/qibo/derivative.py:91]: The usage of error mitigation should be a user choice, unrelated to the calculation of a derivative. We can think to involve this as option into the optimization module, but I would suggest to remove this from the differentiation one.

File: src/qibo/derivative.py [hunk 3]
```diff
@@ -1,7 +1,180 @@
...
+        nshots=10000,
+        noise_model=noise_model,
+        full_output=True,
+        readout={"calibration_matrix": calibration},
+    )
+
+    return optimal_params
+
+
+def calculate_gradients(optimizer, cdr_params):
```
Comment (MatteoRobbiati) [inline src/qibo/derivative.py:107]: The `optimizers` module needs the `derivative` one in some cases (e.g. if the optimization is a gradient descent), but I think the `derivative` should be free from this kind of dependence. 
The idea with the first `parameter_shift` rule was to pass the model (the circuit) as argument and the target parameter index. I understand here you are doing something like that, embedding into the optimizer the model's information. This is ambiguous in some way: ideally we should divide the info (1. model's info, with parameters and so on, 2. optimizer info, with the method chosen to compute gradients if needed, 3. the algorithms used to compute gradients).

I think we can implement something similar we did in the `parameter_shift` but more general, in which the `Parameter` lambda function is in some way encoded into the variational quantum circuit, which can be a `qibo.Circuit` or something more complex (a new qml_model class which can also provide the user with some pre/post-processing tools, etc).

File: src/qibo/derivative.py [hunk 4]
```diff
@@ -11,6 +184,7 @@ def parameter_shift(
     initial_state=None,
     scale_factor=1,
     nshots=None,
+    cdr_params=None,
```
Comment (MatteoRobbiati) [inline src/qibo/derivative.py:187]: I would drop this. I think the mitigation conf can be involved into the optimization module, or in the VQC module if we want to create one.

**Review by MatteoRobbiati — CHANGES_REQUESTED** [review]
Thank you @WanderingMike for this contribution!
I started with reviewing the `derivative.py` file.

Some comments:
1. I think we can be more explicit with the names, in particular I would suggest to rename `_trainablep` as `_variational_params` or something like this;
2. some docstrings are missing in the `Parameter` class, thus I cannot understand some things (e.g. what are the `fixed_params`);
3. this is the main comment, which sums up some of my following suggestions: I think we are mixing some tasks here. We should be able to divide the QML tools we are using here into separate modules. We have the model, the optimizer and the algorithms to compute the derivatives. I suggest we can discuss a new QML model class (e.g. a VariationalQuantumCircuit), which will collect all the Parameters infos, in order to let the `derivative` module free from this kind of dependences. 

Some more suggestions follow.

File: src/qibo/derivative.py [hunk 5]
```diff
@@ -1,7 +1,180 @@
...
 from qibo.hamiltonians.abstract import AbstractHamiltonian
+from qibo.hamiltonians.hamiltonians import SymbolicHamiltonian
+from qibo.models.error_mitigation import CDR, calibration_matrix
+from qibo.symbols import I, Symbol, Z
+
+
+class Parameter:
+    """Object which allows complex gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter"""
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: Solved

File: src/qibo/derivative.py [hunk 6]
```diff
@@ -1,7 +1,180 @@
...
+
+    def __init__(self, func, trainablep, featurep=None):
+        self._trainablep = trainablep
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: Solved.

File: src/qibo/derivative.py [hunk 7]
```diff
@@ -1,7 +1,180 @@
...
+        self._featurep = featurep
+        self.nparams = len(trainablep)
+        self.lambdaf = func
+
+    def _apply_func(self, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: Solved

File: src/qibo/derivative.py [hunk 8]
```diff
@@ -1,7 +1,180 @@
...
+        params = []
+        if self._featurep is not None:
+            params.append(self._featurep)
+        if fixed_params:
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: See Parameter Pull request

File: src/qibo/derivative.py [hunk 9]
```diff
@@ -1,7 +1,180 @@
...
+        return self.lambdaf(*params)
+
+    def _update_params(self, trainablep=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if trainablep:
+            self._trainablep = trainablep
+        if feature:
+            self._featurep = feature
+
+    def get_params(self, trainablep=None, feature=None):
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: See Parameter Pull request

File: src/qibo/derivative.py [hunk 10]
```diff
@@ -1,7 +1,180 @@
...
+        the optimizer's trainable parameter list"""
+        return [start_index + i for i in range(self.nparams)]
+
+    def get_fixed_part(self, trainablep_idx):
+        """Retrieve parameter constant unaffected by a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = self._trainablep[trainablep_idx]
+        return self._apply_func(fixed_params=params)
+
+    def get_scaling_factor(self, trainablep_idx):
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/derivative.py]: See Parameter Pull request

File: src/qibo/derivative.py (hunk 2 above)
Comment (WanderingMike) [reply src/qibo/derivative.py]: I moved this into the optimizer.py file

File: src/qibo/derivative.py (hunk 4 above)
Comment (WanderingMike) [reply src/qibo/derivative.py:89]: Unfortunately the error mitigation can't simply be applied to the end result. If the function $g$ handles our error mitigation, we have: $C * (g(shift_forward) - g(shift_backward)) \noteq g(C * (g(shift_forward) - g(shift_backward))$

File: src/qibo/derivative.py (hunk 3 above)
Comment (WanderingMike) [reply src/qibo/derivative.py]: There exists a draft for a VQC class, however this would be implemented in a separate Pull Request

**Review by MatteoRobbiati — COMMENTED** [review]
Thanks @WanderingMike. A comment follows, which can open a pivotal discussion. I feel if we solve this last bottleneck then we have everything and we can merge the new optimizers interface and proceed with the variational models.

**Comment by scarrazza** [comment]
@MatteoRobbiati we are planning to release qibo later today, shall we postpone this PR for the next release?

**Comment by MatteoRobbiati** [comment]
> @MatteoRobbiati we are planning to release qibo later today, shall we postpone this PR for the next release?

Yes, I think we need to discuss more about this.

File: src/qibo/models/evolution.py [hunk 11]
```diff
@@ -288,9 +288,30 @@ def minimize(self, initial_parameters, method="BFGS", options=None, messages=Fal
             )
 
         args = (self, self.hamiltonian.h1, self.opt_messages, self.opt_history)
-        result, parameters, extra = optimizers.optimize(
-            loss, initial_parameters, args=args, method=method, options=options
-        )
+
+        if method == "sgd":
```
[the remark this answers is not in the document]
Comment (WanderingMike) [reply src/qibo/models/evolution.py:292]: Let's discuss this at the next qibo meeting on 06.12. I think except the loss function being an optional parameter when initialising the optimizer, your suggestion does not substantially change the current code, right?
